### ISSUE TITLE
Map not updating after a while

### ISSUE THREAD

**Issue by Mr-Groch** [PiotrMachowski/Home-Assistant-custom-components-Xiaomi-Cloud-Map-Extractor#41]
I can see lot of:

`Updating xiaomi_cloud_map_extractor camera took longer than the scheduled update interval 0:00:05`

(it's not problem with 5 seconds intervals - same happens when I set 30 seconds)

After HA restart, when vacuum is cleaning, then map is updating properly (even with 5 seconds intervals). But when cleaning ends and vacuum gets back to the dock, `Updating xiaomi_cloud_map_extractor camera took longer than the scheduled update interval 0:00:05` will be back. Next when I start new cleaning, map will be not updated - never. After next HA restart - it will work till next cleaning end...

**Comment by PiotrMachowski** [comment]
It's possible that you have problems with a connection to a vacuum, can you check if you can ping it?

**Comment by Mr-Groch** [comment]
I don't think so - when map isn't updated, I can control it (pause, fan speed change) via xiaomi_miio integration. Vacuum model is Mi Robot Vacuum 1S on cn server.

**Comment by Mr-Groch** [comment]
Each time, HA restart helps. Maybe when updating fails, you should be doing relogin do cloud?

**Comment by PiotrMachowski** [comment]
I also think that this might be a problem, I will try to implement it :)

**Comment by TheRealBadRobot** [comment]
I have same issue.

**Comment by perseus177** [comment]
I have same issue

**Comment by dennishenning** [comment]
I have the same issue

**Comment by LCerebo** [comment]
Same issue here, also tried to increase the timeout with no luck.

**Comment by Dielee** [comment]
Same issue here!

**Comment by jonostanck** [comment]
Same for me unfortunately. Works only for a while and then stops updating. Map is accesible via FloleVac app, and I can Pause/Return to base etc.

**Comment by TheRealBadRobot** [comment]
Is it depends on the main server maybe? My robot connected to Mainland China. Who has this issue wyich server is in use?

**Comment by Dielee** [comment]
My is connected to Mainland China, too!

**Comment by LCerebo** [comment]
> Is it depends on the main server maybe? My robot connected to Mainland China. Who has this issue wyich server is in use?

Mine too! It’s a mijia 1s connected to the Chinese’s servers.

**Comment by Mr-Groch** [comment]
Also Mainland China

**Comment by jonostanck** [comment]
The same here - Mainland China. FloleVac's map feature has it sorted out somehow as it's updating without problems. Might be a re-login thing or... I don't know :)

**Comment by PiotrMachowski** [comment]
I use `de` for my S55 and it works perfectly fine :)

**Comment by TheRealBadRobot** [comment]
> I use `de` for my S55 and it works perfectly fine :)

Which firmware is used?

**Comment by jjusko** [comment]
On my mijia 1s connected to chinese server is the same problem.

**Comment by PiotrMachowski** [comment]
> > I use `de` for my S55 and it works perfectly fine :)
> 
> Which firmware is used?

Old one: 3.3.9_001768

**Comment by TheRealBadRobot** [comment]
> > > I use `de` for my S55 and it works perfectly fine :)
> > 
> > 
> > Which firmware is used?
> 
> Old one: 3.3.9_001768

If I find some way to the downgrade I try with this fw.

**Comment by TheRealBadRobot** [comment]
But before the version 1 it was worked for me perfectly.

**Comment by Squixx** [comment]
Are you running adguard?

I had the same problem in the xiaomi app with the map not loading, or loading very slow:

Looking into ad guard I found 'app-measurement.com' was being blocked each time the map failed to load.

Maybe it's related?

**Comment by dsyntech** [comment]
The problem is in the China region. I changed in Mi Home to the region of Russia and the map is updated after a while.

**Comment by PiotrMachowski** [comment]
Please check if problem still exists in v1.0.2

**Comment by jonostanck** [comment]
Now this happens:

```
Logger: custom_components.xiaomi_cloud_map_extractor.camera
Source: custom_components/xiaomi_cloud_map_extractor/camera.py:179
Integration: xiaomi_cloud_map_extractor (documentation)
First occurred: 23:02:50 (32 occurrences)
Last logged: 23:05:25

Unable log in
```

**Comment by Mr-Groch** [comment]
For me it is better (cn server), I see still lot of:

```
Logger: custom_components.xiaomi_cloud_map_extractor.camera
Source: custom_components/xiaomi_cloud_map_extractor/camera.py:209
Integration: xiaomi_cloud_map_extractor (documentation)
First occurred: 10 stycznia 2021, 11:09:21 (495 occurrences)
Last logged: 11 stycznia 2021, 23:07:32

Unable to retrieve map data
```

In logs, but without HA restarting maps updates for 2 days without bigger problems on each vacuum cleaning

**Comment by Mr-Groch** [comment]
After updating from v1.0.2 to latest - map not updating till HA restart is back again...

**Comment by PiotrMachowski** [comment]
@Mr-Groch does it update after restart? Do you still get `Unable to retrieve map data` in logs?

**Comment by Mr-Groch** [comment]
Yes:

Logger: custom_components.xiaomi_cloud_map_extractor.camera
Source: custom_components/xiaomi_cloud_map_extractor/camera.py:209
Integration: xiaomi_cloud_map_extractor (documentation)
First occurred: 12:21:46 (8 occurrences)
Last logged: 12:28:45

Unable to retrieve map data

**Comment by PiotrMachowski** [comment]
Is map available/updated?

**Comment by Mr-Groch** [comment]
Hmm.. I see that it is not available after HA restart... Maybe there is problem with cn server now, but yesterday my map also wasn't updating (I have updated this component 2 days ago)

**Comment by Mr-Groch** [comment]
Yesterday it looks that it was problem with cn server, but today after some tests on 1.0.2 and 1.0.4 its sure for me that on 1.0.2 map was updating better. In 1.0.4 I see only one occurence of `Unable to retrieve map data` in logs, but map isn't updating (I have observer 1-2 updates per hour, on 1.0.4 I see lot of `Unable to retrieve map data` but map is updating more often

**Comment by kongo09** [comment]
My map also stopped updating. I'm on v 1.0.4 and have no country specified for a server.

**Comment by noorbertt** [comment]
Same issue, after reboot it's working. I don't have pi hole or adguard and I am on global server

**Comment by PiotrMachowski** [comment]
Does it start to work after some time, or is it permanently broken until a restart?

**Comment by jonostanck** [comment]
I also have this issue still - it's permanently broken until restart.

**Comment by kongo09** [comment]
it's now 7 hours later and the map is still broken so I guess it is permanent

**Comment by FeromonDE** [comment]
Same here. it's permanently broken until restart. Roborock S5 Max. DE Server

**Comment by ravensergio** [comment]
Same issue v 1.0.4 on S5 (not MAX), after reboot it's working for a while. I don't a server selected in config.

**Comment by PiotrMachowski** [comment]
How does it work bed9

**Comment by Mr-Groch** [comment]
I've went back to 1.0.2. Lot off `Unable to retrieve map data` in loga but map is updating without need of rebooting HA.

**Comment by PiotrMachowski** [comment]
Can you check out current version from `master` (not released yet)?

**Comment by letsoulfly** [comment]
> Can you check out current version from `master` (not released yet)?

I have tested with `master` version. Seems to be working by checking a few "go-tos" with my S5 Max.

**Comment by Mr-Groch** [comment]
I'm also testing, cn server, no `Unable to retrieve map data` in logs this time, but map isn't updating every 60 second. I think its timeouts with cn server - I see in logs:
```
Logger: homeassistant.helpers.entity
Source: helpers/entity.py:477
First occurred: 10:33:28 (4 occurrences)
Last logged: 11:04:33

Update of camera.mi_robot_vacuum_1s_map is taking over 10 seconds
```
But on 1.0.2 also map wasn't updated every 60 seconds. So far onmaster it is working, but I need to test it little longer to check if it will be working all the time without need for HA restart.

**Comment by PiotrMachowski** [comment]
@Mr-Groch Unfortunately it is hardcoded [here](https://github.com/home-assistant/core/blob/dev/homeassistant/helpers/entity.py#L37)

**Comment by FeromonDE** [comment]
> 
> 
> Can you check out current version from `master` (not released yet)?
I have tested with master version. much better now.

**Comment by PiotrMachowski** [comment]
Fix published in v1.0.5. Can I close this issue now?

**Comment by letsoulfly** [comment]
Updated to v1.0.5... Running latest firmware on my S5 Max (update was 2 days ago, I think), running latest HAss.
Zone cleaning OK. Room cleaning OK. Go-To OK. No bad entries in HAss logging so far. Map/attributes are updating again.

Thank you so much! :)

**Comment by lexural** [comment]
v 1.0.5

```
Logger: homeassistant.components.camera
Source: helpers/entity_platform.py:603
Integration: Камера (documentation, issues)
First occurred: 16:18:25 (61 occurrences)
Last logged: 19:19:04

Updating xiaomi_cloud_map_extractor camera took longer than the scheduled update interval 0:00:05
```
```
2021-01-20 19:28:34 WARNING (MainThread) [homeassistant.components.camera] Updating xiaomi_cloud_map_extractor camera took longer than the scheduled update interval 0:00:05
2021-01-20 19:28:34 WARNING (MainThread) [homeassistant.helpers.entity] Update of camera.xiaomi_cloud_map_extractor is taking over 10 seconds
2021-01-20 19:28:34 WARNING (SyncWorker_1) [custom_components.xiaomi_cloud_map_extractor.camera] Unable to retrieve map data
2021-01-20 19:29:09 WARNING (MainThread) [homeassistant.components.camera] Updating xiaomi_cloud_map_extractor camera took longer than the scheduled update interval 0:00:05
2021-01-20 19:31:09 WARNING (MainThread) [homeassistant.components.camera] Updating xiaomi_cloud_map_extractor camera took longer than the scheduled update interval 0:00:05
2021-01-20 19:31:14 WARNING (MainThread) [homeassistant.helpers.entity] Update of camera.xiaomi_cloud_map_extractor is taking over 10 seconds
2021-01-20 19:31:14 WARNING (MainThread) [homeassistant.components.camera] Updating xiaomi_cloud_map_extractor camera took longer than the scheduled update interval 0:00:05
2021-01-20 19:31:14 WARNING (SyncWorker_10) [custom_components.xiaomi_cloud_map_extractor.camera] Unable to retrieve map data
2021-01-20 19:31:24 WARNING (MainThread) [homeassistant.components.camera] Updating xiaomi_cloud_map_extractor camera took longer than the scheduled update interval 0:00:05
2021-01-20 19:31:29 WARNING (MainThread) [homeassistant.helpers.entity] Update of camera.xiaomi_cloud_map_extractor is taking over 10 seconds
2021-01-20 19:31:29 WARNING (MainThread) [homeassistant.components.camera] Updating xiaomi_cloud_map_extractor camera took longer than the scheduled update interval 0:00:05
```

**Comment by jonostanck** [comment]
After several hours it still updates - yay! Thank you for this update, seems the issue is resolved. :)

**Comment by PiotrMachowski** [comment]
@lexural Does it update despite errors?

**Comment by lexural** [comment]
> @lexural Does it update despite errors?

Something has stopped updating and loading the map altogether.

**Comment by FeromonDE** [comment]
```


Logger: homeassistant.helpers.entity
Source: custom_components/xiaomi_cloud_map_extractor/xiaomi_cloud_connector.py:125
First occurred: 13:23:34 (2 occurrences)
Last logged: 13:23:44
Update for camera.roborock_s5_max fails

Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 278, in async_update_ha_state
    await self.async_device_update()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 474, in async_device_update
    raise exc
  File "/usr/local/lib/python3.8/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/config/custom_components/xiaomi_cloud_map_extractor/camera.py", line 200, in update
    map_data = self._connector.get_map(self._country, map_name, self._colors, self._drawables, self._texts,
  File "/config/custom_components/xiaomi_cloud_map_extractor/xiaomi_cloud_connector.py", line 130, in get_map
    response = self.get_raw_map_data(country, map_name)
  File "/config/custom_components/xiaomi_cloud_map_extractor/xiaomi_cloud_connector.py", line 141, in get_raw_map_data
    map_url = self.get_map_url(country, map_name)
  File "/config/custom_components/xiaomi_cloud_map_extractor/xiaomi_cloud_connector.py", line 125, in get_map_url
    if api_response is None or "result" not in api_response or "url" not in api_response["result"]:
TypeError: argument of type 'NoneType' is not iterable
```
Version 1.0.5

**Comment by PiotrMachowski** [comment]
@lexural So the map did update?

@FeromonDE Does the map update despite error? Do these errors appear all the time or occasionally?

**Comment by FeromonDE** [comment]
Yes. Occasionaly.

**Comment by VirtualL** [comment]
+1 for this issue

**Comment by TTLucian** [comment]
I also have this issue. Map is updating but with delays and some elements missing.
Errors:
```
[homeassistant.components.camera] Updating xiaomi_cloud_map_extractor camera took longer than the scheduled update interval 0:00:05
[custom_components.xiaomi_cloud_map_extractor.camera] Unable to retrieve map data
```
Also camera component shows this:
```
[homeassistant.components.camera] The websocket command 'camera_thumbnail' has been deprecated
```

Also, map looks like this in Home Assistant:

![ha](https://user-images.githubusercontent.com/1855276/106413605-2cb42a00-6453-11eb-8483-f6d273e04367.jpg)


And it looks like this in Xiaomi app:

![x](https://user-images.githubusercontent.com/1855276/106413619-389fec00-6453-11eb-9002-1d4c7941a01e.jpg)

**Comment by lexural** [comment]
Logger: homeassistant.helpers.entity
Source: custom_components/xiaomi_cloud_map_extractor/xiaomi_cloud_connector.py:125
First occurred: 10:05:37 (1 occurrences)
Last logged: 10:05:37

Update for camera.xiaomi_cloud_map_extractor fails
```
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 278, in async_update_ha_state
    await self.async_device_update()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 474, in async_device_update
    raise exc
  File "/usr/local/lib/python3.8/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/config/custom_components/xiaomi_cloud_map_extractor/camera.py", line 200, in update
    map_data = self._connector.get_map(self._country, map_name, self._colors, self._drawables, self._texts,
  File "/config/custom_components/xiaomi_cloud_map_extractor/xiaomi_cloud_connector.py", line 130, in get_map
    response = self.get_raw_map_data(country, map_name)
  File "/config/custom_components/xiaomi_cloud_map_extractor/xiaomi_cloud_connector.py", line 141, in get_raw_map_data
    map_url = self.get_map_url(country, map_name)
  File "/config/custom_components/xiaomi_cloud_map_extractor/xiaomi_cloud_connector.py", line 125, in get_map_url
    if api_response is None or "result" not in api_response or "url" not in api_response["result"]:
TypeError: argument of type 'NoneType' is not iterable
```

**Comment by PiotrMachowski** [comment]
@FeromonDE @lexural 
I have created a separate issue for this error: #78 

@TTLucian
Delays in map updates are rather a problem with cloud connection - some of them are probably blocked by cloud.

An issue for `camera_thumbnail` problem is [here](https://github.com/PiotrMachowski/lovelace-xiaomi-vacuum-map-card/issues/67).

Your map seems fine to me, if you want the same colors you can play around with [colors configuration](https://github.com/PiotrMachowski/Home-Assistant-custom-components-Xiaomi-Cloud-Map-Extractor#colors-configuration) (to remove grey areas you can edit `color_map_wall_v2` and `color_grey_wall`) and [room colors configuration](https://github.com/PiotrMachowski/Home-Assistant-custom-components-Xiaomi-Cloud-Map-Extractor#room-colors-configuration).

**Comment by PiotrMachowski** [comment]
It seems that problems with map not being updated are now gone, so I will close this issue.

**Comment by LCerebo** [comment]
A side note to this issue.
I know it is a rather unusual case, but if you are running a tor node on the same ip of your home-assistant instance you will be banned from the Chinese servers and get the error reported in this issue.
You'll also experience connection problems through the Xiaomi app.

**Comment by combative-american** [comment]
Just finished setting up the integration, no map image is displayed. I do use the Mail and Packages Integration which downloads the .gif image to my /config/www/... folder. Is this integration supposed to work the same way? I looked through everything and did not see any images. The only errors I received were the Updating xiaomi_cloud_map_extractor camera took longer than the scheduled update interval 0:00:05.

**Comment by ShadowJonathan** [comment]
I'm having the same problem, the map updates the first time the bot leaves the charger, but then doesn't update at all until it starts charging again. (z10 Pro)

**Comment by PiotrMachowski** [comment]
@ShadowJonathan https://github.com/PiotrMachowski/Home-Assistant-custom-components-Xiaomi-Cloud-Map-Extractor/issues/299
